"Spiral" annular mesh generation

leorachapuis opened the pull request:
This implements a new mesh derived from MooseMesh, called "SpiralAnnularMesh".
It has an annular shape and triangular elements.

Refs #12205.

friedmud committed:
Create stub SpiralAnnularMesh to get started refs #12205

leorachapuis committed:
Create a new mesh derived from MooseMesh.

It's called SpiralAnnularMesh and allows to create a mesh with an
annular shape and triangular elements.  Also create the test files for
the new mesh SpiralAnnularMesh.

Refs #12205.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/238310/) on ace82de wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12220/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12220/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1e75d719f1474b7f2031b36349f20662be8efc66`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/238355/) on 46c8821 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12220/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12220/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b5b05406368106f1de256fa5f0bde99c1479545b`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/238371/) on c23f78e wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12220/site)

This comment will be updated on new commits.

jwpeterson commented:
@leorachapuis you'll need to add some markdown documentation for the new class to get the Documentation test target to pass:
```
MooseDocs.base.renderers: 
ERROR:The following errors were reported when accessing the '/Mesh' syntax.
Failed to locate a page with the given filename: mesh/SpiralAnnularMesh.md
```

jwpeterson commented on framework/include/mesh/SpiralAnnularMesh.h:
```diff
@@ -0,0 +1,69 @@
+  unsigned int _num_rings;
+
+  /// The boundary id to use for the cylinder.
+  const boundary_id_type &_cylinder_bid, _exterior_bid;
```
Oh, I meant to mention when I saw this line of code earlier: the way that reference variable declarations work means that only the first of these (`_cylinder_bid`) is actually a reference type. The second (`_exterior_bid`) is a normal `boundary_id_type`. I think in this case there's no reason to make them references at all, so simply dropping the ampersand should be sufficient
```
const boundary_id_type _cylinder_bid, _exterior_bid;
```

jwpeterson commented on framework/src/mesh/SpiralAnnularMesh.C:
```diff
@@ -0,0 +1,306 @@
+  params.addRequiredRangeCheckedParam<unsigned int>(
+      "nodes_per_ring", "nodes_per_ring>0", "Number of nodes on each ring.");
+  params.addParam<bool>(
+      "my_second_order", false, "Generate mesh of TRI6 elements instead of TRI3 elements.");
```
The plan is to eventually move the `second_order` parameter out of `SetupMeshAction` and into `MooseMesh` itself. In the meantime, this is a workaround that avoids our param clashing with the one in the Action.

jwpeterson commented on framework/src/mesh/SpiralAnnularMesh.C:
```diff
@@ -0,0 +1,306 @@
+  if (_outer_radius <= _inner_radius)
+    mooseError("SpiralAnnularMesh: outer_radius must be greater than inner_radius");
+  if (_num_rings == 0)
+    mooseError("Cannot create mesh with _num_rings == 0");
```
@leorachapuis I also have not tested this code with an odd-numbered `_nodes_per_ring` parameter. I don't see any reason why it wouldn't work, but it would still be good to test...

jwpeterson replied:
The same comment applies to the `_num_rings` parameter: I believe I have only tested with even numbers of rings, but an odd number should also be possible.

jwpeterson commented on framework/src/mesh/SpiralAnnularMesh.C:
```diff
@@ -0,0 +1,306 @@
+    Real f, df;
+    newton(f, df, alpha);
+
+    while (std::abs(f) > 1.e-9)
```
It's possible that this Newton iteration could fail to converge (thus making this an infinite while loop). This has not happened yet in my testing but we should guard against it by specifying a maximum number of allowed iterations (say 25) and if that is exceeded throw a `mooseError`.

jwpeterson commented on framework/src/mesh/SpiralAnnularMesh.C:
```diff
@@ -0,0 +1,306 @@
+    }
+  }
+
+  mesh.write("bassi_rebay_mesh.e");
```
This line needs to be removed! We wrote the mesh in the original test code but don't want to do that here.

leorachapuis committed:
Add documentation for SpiralAnnularMesh.
Also fix the issues requested.
Refs #12205.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/238631/) on 3266099 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12220/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12220/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b5b05406368106f1de256fa5f0bde99c1479545b`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 05aed3d48770448843f0beec15f4ed11beadf932
+Subproject commit cfc475ff199568581219e11d48b90f5b9aa15310
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

jwpeterson commented:
Ugh, looks like clang-format still doesn't like one blank line :_( This can sometimes happen if you use a different version of clang on your laptop than is being used by CIVET...

```
----------------------- Start diff -----------------
diff --git a/framework/src/mesh/SpiralAnnularMesh.C b/framework/src/mesh/SpiralAnnularMesh.C
index e432b9b..8a65d48 100644
--- a/framework/src/mesh/SpiralAnnularMesh.C
+++ b/framework/src/mesh/SpiralAnnularMesh.C
@@ -307,5 +307,4 @@ SpiralAnnularMesh::buildMesh()
       elem->point(nos[2]) = Point(new_r * std::cos(new_theta), new_r * std::sin(new_theta), 0.);
     }
   }
-
 }
----------------------- End diff -----------------
```

jwpeterson commented:
@leorachapuis there are a few missing documentation things required for the tests in this directory, I will take a crack at adding them.

jwpeterson committed:
Add requirement/design/issues for mesh_generation tests.

jwpeterson committed:
Improve range checks on parameters.

* After testing with smaller numbers of nodes/per ring, I found that 6
  is the minimum number where you can get a sensible result.
* Although having 1 ring would be technically possible, it does not
  work with our Newton iteration that computes the radial bias, so we're
  restricting this to parameter to be >= 2.
* Rename the parameter for setting quadratic elements to
  "use_tri6". This will eventually go away once we move the
  "second_order" parameter out of SetupMeshAction.
* Remove redundant check on num_rings parameter, it is already range-checked.

Refs #12205.

jwpeterson commented:
I ran a few tests locally and confirmed that odd numbers of nodes/ring work fine. I did need to tweak some of the other range check, but other than that, this should be good to go.
